### PR TITLE
Add a dynamic iterator for domains

### PR DESCRIPTION
**Pull request by Ldelaney**
Adds to DynamicIters a dynamic iterator for domains, which invokes the existing dynamic iterator for ranges. Adds a correctness test (to /functions/iterators, though it is possible a /modules/standard/DynamicIters directory should be made and it should be put there instead) to confirm that the new iterator leads and follows properly. Adds a version of mandelbrot-unrolled using the new iterator, as well as a new graph to compare its performance with the other iterator. (It is possible that this should not have its own graph and should instead be part of mandelbrot-all).

Just running tests on my work machine, all of the release tests succeed, as do all of /functions/iterators and all of /studies/shootout/mandelbrot.

### CODE REVIEW

**Comment by Ldelaney** [comment]
@benharsh would you review this?

File: modules/standard/DynamicIters.chpl [hunk 1]
```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
   }
 }
 
+//************************* Multidimensional Dynamic iterator
+/*
+
+  :arg c: The domain to iterate over. The rank of the domain must be greater 
+          than one. If using a 1d domain, the user should use the existing
+          dynamic iterators (use DynamicIters).
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: Is this a technical limitation? If not, I think we would want to support 1D domains for completeness. We still find users writing code like this at times:

```chpl
forall i dynamic({1..10}, 1) { ... }
```

instead of passing in a range directly. It's not the style we suggest, but if we can handle it I think that's better than a compile-time error.

File: modules/standard/DynamicIters.chpl [hunk 2]
```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
...
+  :arg numTasks: The number of tasks to use. Must be >= zero. If this argument
+                 has the value 0, it will use the value indicated by
+                 ``dataParTasksPerLocale``.
+  :type numTasks: `int`
+
+  :arg parDim: The index of the dimension to parallelize across. Must be > 0.
+                Must be <= the rank of the domain ``c``. Defaults to 1. 
+  :type parDim: `int`
+
+  :yields: Slices of the domain ``c``
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: I think it should be "Indices" instead of "Slices".

File: modules/standard/DynamicIters.chpl [hunk 3]
```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
...
+  Given an input domain ``c``, each task is assigned slices of ``c``. The
+  chunks each have ``chunkSize`` slices in them (or the remaining iterations
+   if there are fewer than ``chunkSize``). This continues until there are no 
+   remaining iterations in the dimension of ``c`` indicated by ``parDim``. 
+
+  LIMITATIONS:
+  user can select parDim, but only if they also define numTasks (even to 0)
+    should parDim be a non-default param?
+  explicitly types the ranges (so they will likely only work with ints)
+  May or may not properly lead/follow with other domain iterators
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: Two things:

You can name the argument 'parDim' to give it a value, like this:

```chpl
var dom = {1..5, 1..10};
forall i in dynamic(dom, 1, parDIm=2) do ...
```

Second, I think by the end of this you should be able to say that this iterator should lead and follow with rectangular domains correctly, and with any domains that interact nicely with rectangular domains.

File: modules/standard/DynamicIters.chpl [hunk 4]
```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
...
+*/
+
+//This is the serial version of this iterator
+iter dynamic(c:domain, chunkSize:int, numTasks:int=0, parDim:int=1) 
+{
+  if debugDynamicIters then
+    writeln("Serial Multidimensional Dynamic Iterator, working with domain: ", c);
+  
+  var yieldTuple :c.rank * int;
+  for yieldTuple in c do yield yieldTuple;
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: The first ``yieldTuple`` here is a separate variable from the ``yieldTuple`` in the loop, so it doesn't quite work like C/C++ in this case. You could make that first ``yieldTuple`` a string and everything would still work.

File: modules/standard/DynamicIters.chpl [hunk 5]
```diff
@@ -154,6 +154,113 @@ where tag == iterKind.follower
...
+}
+
+
+//this is the leader
+iter dynamic(param tag:iterKind, c:domain, chunkSize:int, numTasks:int=0, parDim : int = 1) 
+  where tag == iterKind.leader 
+  {
+    //caller's responsibility to use a valid chunk size
+    assert(chunkSize > 0);
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: I think these asserts should have a message passed in, like this:

```chpl
assert(chunkSize > 0, " Chunk size must be greater than zero.");
```

File: modules/standard/DynamicIters.chpl (hunk 3 above)
Comment (Ldelaney) [inline modules/standard/DynamicIters.chpl]: I can add a call to the range iterator to handle one dimensional domains.

I used slices instead of indices because if the function is passed the domain {1..10, 3..4, 5..6} with chunk size one, it will return (1..1, 3..4, 5..6), (2..2, 3..4, 5..6), etc., and  I wanted to be clear that it is yielding a tuple of ranges. I also was originally returning a subdomain instead of a tuple of indices because I was confused about how leader/follower iterators for domains worked.

I have a test for leading/following with the default domain iterator, but don’t know what further testing is needed to prove that it works correctly with all of the rectangular domain iterators all of the time.

File: modules/standard/DynamicIters.chpl (hunk 5 above)
Comment (Ldelaney) [inline modules/standard/DynamicIters.chpl]: should that be changed for the existing dynamic iterators with similar asserts as well?

Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: That would be good, but we could do that in a separate PR.

File: test/functions/iterators/ldelaney/zipDynamic3d.chpl [hunk 6]
```diff
@@ -0,0 +1,55 @@
+use DynamicIters;
```
Comment (benharsh) [inline test/functions/iterators/ldelaney/zipDynamic3d.chpl]: A concern I have with this test is that it wouldn't detect a bug where we accidentally yielded the same index twice. I don't think that would happen today, but if we can write this test in a way that would catch such an issue, that would be preferable.

I think another way to write this could be to declare an array of integers over the desired domain, and then in the loop body increment each index by one. After the loop, you can assert that each index was visited once.

File: test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  const ydim = 0..n-1;
+  const xdim = 0..((n+7)/8)-1;
+  var pixels : [ydim][xdim] uint(8);
+
+  forall i in ydim {
+    Cib[i] = i * invN - 1.0;
+    Crb[i] = i * invN - 1.5;
+  }
+
+  forall xy in dynamic({xdim, ydim}, chunkSize, 0, 2){
```
Comment (benharsh) [inline test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl]: You don't need to change this, but in case you weren't aware you can write a forall like this to do the tuple expansion for you:

```chpl
forall (x,y) in dynamic(...)
```

File: test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS [hunk 8]
```diff
@@ -0,0 +1 @@
+--n=16000
```
Comment (benharsh) [inline test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS]: I think this is a good correctness test, but I don't think it's what we want for a performance test. I think the mandelbrot test you added will be a much better for comparison.

File: modules/standard/DynamicIters.chpl [hunk 9]
```diff
@@ -154,6 +154,111 @@ where tag == iterKind.follower
...
+  :arg parDim: The index of the dimension to parallelize across. Must be > 0.
+                Must be <= the rank of the domain ``c``. Defaults to 1. 
+  :type parDim: `int`
+
+  :yields: Indices of the domain ``c`` as a tuple of ranges
+
+  Given an input domain ``c``, each task is assigned slices of ``c``. The
+  chunks each have ``chunkSize`` slices in them (or the remaining iterations
+   if there are fewer than ``chunkSize``). This continues until there are no 
+   remaining iterations in the dimension of ``c`` indicated by ``parDim``.
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: This is a little silly, but our documentation tool (sphinx) complains here about indentation. I think you just need to shift these two lines a little to the left to line up correctly. You can look for errors by running 'make docs' from $CHPL_HOME.

File: test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS (hunk 8 above)
Comment (Ldelaney) [inline test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS]: This comment is on the performance execution options for the mandelbrot test. Is it correct to assume that you are talking about the zipDynamic3d test?

File: test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl (hunk 7 above)
Comment (Ldelaney) [inline test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl]: would it change the performance at all?

File: test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS (hunk 8 above)
Comment (benharsh) [inline test/studies/shootout/mandelbrot/ldelaney/PERFEXECOPTS]: Oops, you're right, nevermind!

File: test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl (hunk 7 above)
Comment (benharsh) [inline test/studies/shootout/mandelbrot/ldelaney/mandelbrot-unrolled-DynamicNDIter.chpl]: I'd be surprised if it did, but you could test it to be sure.

File: modules/standard/DynamicIters.chpl [hunk 10]
```diff
@@ -154,6 +154,93 @@ where tag == iterKind.follower
...
+  if debugDynamicIters then
+    writeln("Serial Multidimensional Dynamic Iterator, working with domain: ", c);
+  
+
+  for yieldTuple in c do yield yieldTuple;
+}
+
+
+//this is the leader
+iter dynamic(param tag:iterKind, c:domain, chunkSize:int, numTasks:int=0, parDim : int = 1)
```
Comment (benharsh) [inline modules/standard/DynamicIters.chpl]: I should have caught this earlier, but I think we want a "no doc" pragma on the leader/follower iterators, like the range versions.

File: test/functions/iterators/ldelaney/dynamicIterStrideable.chpl [hunk 11]
```diff
@@ -0,0 +1,18 @@
+use DynamicIters;
+proc main(){
+	var r1 = 1..10 by 2;
+	var r2 = 7..8;
+	var r3 = -2..-1;
+	var d1 = {r1, r2, r3};
+	var bigArr : [r1][r2][r3] int;
```
Comment (benharsh) [inline test/functions/iterators/ldelaney/dynamicIterStrideable.chpl]: You don't need to change this, but for future reference you can write this as:

```chpl
var bigArr : [d1] int;
```

To get a multidimensional array.

**Comment by benharsh** [comment]
The rest looks good! Feel free to merge in after the "no doc" change.
